src/metabase/query_processor/parameters.clj

```diff
@@ -0,0 +1,248 @@
+        "thisyear"   ["TIME_INTERVAL" field "current" "year"]
+        "lastweek"   ["TIME_INTERVAL" field "last" "week"]
+        "lastmonth"  ["TIME_INTERVAL" field "last" "month"]
+        "lastyear"   ["TIME_INTERVAL" field "last" "year"]
+        "yesterday"  ["=" field ["relative_datetime" -1 "day"]]
+        "today"      ["=" field ["relative_datetime" "current"]]))))
+
+(defn- merge-filter-clauses [base addtl]
+  (cond
+    (and (seq base)
+         (seq addtl)) ["AND" base addtl]
+    (seq base)        base
+    (seq addtl)       addtl
+    :else             []))
+
+(defn- expand-params-mbql [query-dict [{:keys [target value], :as param} & rest]]
+  (if param
+    (if (and param target value)
+      (let [filter-subclause (build-filter-clause param)
+            query            (assoc-in query-dict [:query :filter] (merge-filter-clauses (get-in query-dict [:query :filter]) filter-subclause))]
+        (expand-params-mbql query rest))
+      (expand-params-mbql query-dict rest))
+    query-dict))
+
+
+;;; +-------------------------------------------------------------------------------------------------------+
+;;; |                                           SQL QUERIES                                             |
+;;; +-------------------------------------------------------------------------------------------------------+
+
+
+(defn- extract-dates [value report-timezone]
+  (if-not (contains? relative-dates value)
+    ;; absolute date range
+    (absolute-date->range value)
+    ;; relative date range
+    (relative-date->range value report-timezone)))
+
+(defn- expand-date-range-param [report-timezone {[target param-name] :target, param-type :type, param-value :value, :as param}]
+  (if-not (= param-type "date")
+    param
+    (let [{:keys [start end]} (extract-dates param-value report-timezone)]
+      [(assoc param :target [target (str param-name ":start")], :value start)
+       (assoc param :target [target (str param-name ":end")],   :value end)])))
+
+(defn- substitute-param [param-name value query]
+  ;; TODO: escaping and protection against SQL injection!
+  (s/replace query (re-pattern (str "\\{\\{" param-name "\\}\\}")) value))
+
+(defn- substitute-all-params [query-dict [{:keys [value], [_ param-name] :target, :as param} & rest]]
+  (if param
+    (if-not (and param param-name value (string? param-name))
+      (substitute-all-params query-dict rest)
+      (let [query (update-in query-dict [:native :query] (partial substitute-param param-name value))]
+        (substitute-all-params query rest)))
+    query-dict))
+
+(def ^:private ^:const outer-clause-pattern #"\[\[.*?\]\]")
+(def ^:private ^:const outer-clause-prefix-pattern #"^\[\[(.*?)\s.*\]\]$")
+(def ^:private ^:const incomplete-outer-clause-pattern #"\[\[.*?\{\{.*?\}\}.*?\]\]")
+(def ^:private ^:const inner-clause-pattern #"<(.*?)>")
+(def ^:private ^:const incomplete-inner-clause-pattern #"<.*?\{\{.*?\}\}.*?>")
```

Comment:
It would be nice to see these constants [indented together](https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#align-single-line-vars)

And it would be extremely [useful](https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#judiciously-use-comments) to see some quick blurbs about the purpose of these functions. I or any potential contributors won't have any idea what they're for or what they're supposed to do.